### PR TITLE
Feature/programming exercise/mock connector requests for tests

### PR DESCRIPTION
**Pull request by ungaralex**
### Checklist
- [x] I tested *all* changes and *all* related features with different users (student, tutor, instructor, admin) on the test server https://artemistest.ase.in.tum.de.
- [x] Server: I added multiple integration tests (Spring) related to the features
- [x] Server: I documented the Java code using JavaDoc style.

### Motivation and Context

Currently, we always mock our connector services (e.g. Bitbucket, Bamboo) when testing programming exercises. We want to only mock the requests to the CI and VCS servers and not the Spring service itself, so we also cover the code in these cases. 

### Description

I added two new tests (setup and start programming exercise), for  which all requests to external services get mocked by Spring's `MockRestServiceServer`. The added benefit, besides that we actually run the code in our `BambooService` and `BitbucketService` is that we also test the actual values that requests to the external services contain. A request gets mocked by matching the contents and headers against some predefined values. Only when these values are equal, the request gets mocked. Otherwise, an exception is thrown. Because of that, the test will only succeed if the connector services process the parameters as expected.

**Because this PR would be too big, I made a second PR containing only the production code changes: #1169 . After we merge this, I will finalize this PR.**
In order to be able to mock everything, I also had to make a couple of other adaptions:
- I added a `BambooServerConfiguration` that creates beans for the Bamboo server and client. Before that, these objects got instantiated in the methods themselves, so we couldn't mock anything that got routed through them. Since we send requests through these objects, we also have to be able to inject mock instances, so I had to refactor the configuration into Spring beans.
- The `BitbucketBambooUpdateService` is not new! I just refactored the old `BambooUpdateService` that only encapsulated the `BitbucketBambooUpdateService` in an inner class. This made no sense and meant only having one additional unused object.
- The Bitbucket and Bamboo services now have dedicated `RestTemplates` that are necessary for mocking the REST requests for both of them separately . Those templates still are plain Spring `RestTemplates` without any modification.
- The `BambooBuildPlanUpdateProvider` is **unmodified** code from the old `BitbucketBambooUpdateService` that @krusche once extracted from the third party `BambooClient`. I had to move this code to a separate bean, because it also calls services, that perform REST requests (but not using Spring's RestTemplate). That is why we have to mock this whole service when executing tests, which is only possible if it is an injected bean.
- There is a couple of new DTOs that we send to the Bitbucket/Bamboo servers. In our old code,  we just manually put those values into Maps and serialized them. Testing this is not ideal, because we would have to copy a lot of code. This way, all models get managed in those DTOs and we can be sure, that both the tests and the productive code use the same data structures (So testing for equality is easier).

In this PR:
- I also started moving String values for our REST API endpoints to constant classes, so we don't have to copy them in our tests, but can just reference the constant, so we can guarantee that the productive and the test code both use the same values.

### Steps for Testing

1. Run all Java tests
2. Create, edit and participate in a programming exercise. Everything should work as before.

### CODE REVIEW

**Comment by ungaralex** [comment]
@krusche let me know if you want me to keep the changes to the API endpoint Strings (new `Endpoints` constant classes) or If I should revert them.

File: src/test/java/de/tum/in/www1/artemis/util/DatabaseUtilService.java [hunk 1]
```diff
@@ -636,6 +662,7 @@ public ModelingSubmission addModelingSubmission(ModelingExercise exercise, Model
         return submission;
     }
 
+    @Transactional  // needed for adding values to persistent hash set (submissions)
```
Comment (krusche) [inline src/test/java/de/tum/in/www1/artemis/util/DatabaseUtilService.java]: can we avoid using transactional here? I removed it in a previous commit. I think you can just save the objects into the repositories if needed.

File: src/main/java/de/tum/in/www1/artemis/web/rest/ProgrammingExerciseResource.java [hunk 2]
```diff
@@ -789,10 +791,42 @@ public ProgrammingExerciseResource(ProgrammingExerciseRepository programmingExer
...
+
+        public static final String SETUP = PROGRAMMING_EXERCISES + "/setup";
+
+        public static final String GET_FOR_COURSE = "/courses/{courseId}/programming-exercises";
+
+        public static final String IMPORT = PROGRAMMING_EXERCISES + "/import/{sourceExerciseId}";
+
+        public static final String PROBLEM = "/programming-exercises-problem";
+
+        public static final String PROGRAMMING_EXERCISE = PROGRAMMING_EXERCISES + "/{id}";
```
Comment (krusche) [inline src/main/java/de/tum/in/www1/artemis/web/rest/ProgrammingExerciseResource.java]:
```suggestion
        public static final String PROGRAMMING_EXERCISE = PROGRAMMING_EXERCISES + "/{exerciseId}";
```

File: src/main/java/de/tum/in/www1/artemis/web/rest/ProgrammingExerciseResource.java [hunk 3]
```diff
@@ -764,9 +766,9 @@ public ProgrammingExerciseResource(ProgrammingExerciseRepository programmingExer
      * @param exerciseId the id of a ProgrammingExercise
      * @return the ResponseEntity with status 200 (OK) and ProgrammingExerciseTestCaseStateDTO. Returns 404 (notFound) if the exercise does not exist.
      */
-    @GetMapping(value = "/programming-exercises/{exerciseId}/test-case-state")
+    @GetMapping(Endpoints.TEST_CASE_STATE)
     @PreAuthorize("hasAnyRole('TA', 'INSTRUCTOR', 'ADMIN')")
-    public ResponseEntity<ProgrammingExerciseTestCaseStateDTO> hasAtLeastOneStudentResult(@PathVariable Long exerciseId) {
+    public ResponseEntity<ProgrammingExerciseTestCaseStateDTO> hasAtLeastOneStudentResult(@PathVariable("id") Long exerciseId) {
```
Comment (krusche) [inline src/main/java/de/tum/in/www1/artemis/web/rest/ProgrammingExerciseResource.java]:
```suggestion
    public ResponseEntity<ProgrammingExerciseTestCaseStateDTO> hasAtLeastOneStudentResult(@PathVariable("exerciseId") Long exerciseId) {
```

File: src/main/java/de/tum/in/www1/artemis/web/rest/ProgrammingExerciseResource.java [hunk 4]
```diff
@@ -697,7 +699,7 @@ public ProgrammingExerciseResource(ProgrammingExerciseRepository programmingExer
      * @param id The ID of the programming exercise for which the structure oracle should get generated
      * @return The ResponseEntity with status 201 (Created) or with status 400 (Bad Request) if the parameters are invalid
      */
-    @GetMapping(value = "/programming-exercises/{id}/generate-tests", produces = MediaType.TEXT_PLAIN_VALUE)
+    @GetMapping(value = Endpoints.GENERATE_TESTS, produces = MediaType.TEXT_PLAIN_VALUE)
     @PreAuthorize("hasAnyRole('INSTRUCTOR', 'ADMIN')")
     @FeatureToggle(Feature.PROGRAMMING_EXERCISES)
     public ResponseEntity<String> generateStructureOracleForExercise(@PathVariable Long id) {
```
Comment (krusche) [inline src/main/java/de/tum/in/www1/artemis/web/rest/ProgrammingExerciseResource.java]:
```suggestion
    public ResponseEntity<String> generateStructureOracleForExercise(@PathVariable Long exerciseId) {
```

File: src/main/java/de/tum/in/www1/artemis/web/rest/ProgrammingExerciseResource.java [hunk 5]
```diff
@@ -591,10 +593,10 @@ public ProgrammingExerciseResource(ProgrammingExerciseRepository programmingExer
      * @return ResponseEntity with status
      * @throws IOException if submissions can't be zippedRequestBody
      */
-    @PostMapping(value = "/programming-exercises/{exerciseId}/export-repos-by-student-logins/{studentIds}")
+    @PostMapping(Endpoints.EXPORT_SUBMISSIONS_BY_STUDENT)
     @PreAuthorize("hasAnyRole('TA', 'INSTRUCTOR', 'ADMIN')")
     @FeatureToggle(Feature.PROGRAMMING_EXERCISES)
-    public ResponseEntity<Resource> exportSubmissionsByStudentLogins(@PathVariable Long exerciseId, @PathVariable String studentIds,
+    public ResponseEntity<Resource> exportSubmissionsByStudentLogins(@PathVariable("id") Long exerciseId, @PathVariable String studentIds,
```
Comment (krusche) [inline src/main/java/de/tum/in/www1/artemis/web/rest/ProgrammingExerciseResource.java]:
```suggestion
    public ResponseEntity<Resource> exportSubmissionsByStudentLogins(@PathVariable("exerciseId") Long exerciseId, @PathVariable String studentIds,
```

File: src/main/java/de/tum/in/www1/artemis/web/rest/ProgrammingExerciseResource.java [hunk 6]
```diff
@@ -646,10 +648,10 @@ public ProgrammingExerciseResource(ProgrammingExerciseRepository programmingExer
      * @return ResponseEntity with status
      * @throws IOException if submissions can't be zippedRequestBody
      */
-    @PostMapping(value = "/programming-exercises/{exerciseId}/export-repos-by-participation-ids/{participationIds}")
+    @PostMapping(Endpoints.EXPORT_SUBMISSIONS_BY_PARTICIPATIONS)
     @PreAuthorize("hasAnyRole('TA', 'INSTRUCTOR', 'ADMIN')")
     @FeatureToggle(Feature.PROGRAMMING_EXERCISES)
-    public ResponseEntity<Resource> exportSubmissionsByParticipationIds(@PathVariable Long exerciseId, @PathVariable String participationIds,
+    public ResponseEntity<Resource> exportSubmissionsByParticipationIds(@PathVariable("id") Long exerciseId, @PathVariable String participationIds,
```
Comment (krusche) [inline src/main/java/de/tum/in/www1/artemis/web/rest/ProgrammingExerciseResource.java]:
```suggestion
    public ResponseEntity<Resource> exportSubmissionsByParticipationIds(@PathVariable("exerciseId") Long exerciseId, @PathVariable String participationIds,
```

File: src/main/java/de/tum/in/www1/artemis/web/rest/ProgrammingExerciseResource.java [hunk 7]
```diff
@@ -554,7 +556,7 @@ public ProgrammingExerciseResource(ProgrammingExerciseRepository programmingExer
      *              403 (Forbidden) if the user is not admin and course instructor or
      *              500 (Internal Server Error)
      */
-    @PutMapping(value = "/programming-exercises/{id}/combine-template-commits", produces = MediaType.TEXT_PLAIN_VALUE)
+    @PutMapping(value = Endpoints.COMBINE_COMMITS, produces = MediaType.TEXT_PLAIN_VALUE)
     @PreAuthorize("hasAnyRole('INSTRUCTOR', 'ADMIN')")
     @FeatureToggle(Feature.PROGRAMMING_EXERCISES)
     public ResponseEntity<Void> combineTemplateRepositoryCommits(@PathVariable Long id) {
```
Comment (krusche) [inline src/main/java/de/tum/in/www1/artemis/web/rest/ProgrammingExerciseResource.java]:
```suggestion
    public ResponseEntity<Void> combineTemplateRepositoryCommits(@PathVariable Long exerciseId) {
```

File: src/main/java/de/tum/in/www1/artemis/web/rest/ProgrammingExerciseResource.java [hunk 8]
```diff
@@ -489,7 +491,7 @@ public ProgrammingExerciseResource(ProgrammingExerciseRepository programmingExer
      * @param id the id of the programmingExercise to retrieve
      * @return the ResponseEntity with status 200 (OK) and with body the programmingExercise, or with status 404 (Not Found)
      */
-    @GetMapping("/programming-exercises/{id}/with-participations")
+    @GetMapping(Endpoints.PROGRAMMING_EXERCISE_WITH_PARTICIPATIONS)
     @PreAuthorize("hasAnyRole('TA', 'INSTRUCTOR', 'ADMIN')")
     public ResponseEntity<ProgrammingExercise> getProgrammingExerciseWithAllParticipations(@PathVariable Long id) {
```
Comment (krusche) [inline src/main/java/de/tum/in/www1/artemis/web/rest/ProgrammingExerciseResource.java]:
```suggestion
    public ResponseEntity<ProgrammingExercise> getProgrammingExerciseWithAllParticipations(@PathVariable Long exerciseId) {
```

File: src/main/java/de/tum/in/www1/artemis/web/rest/ProgrammingExerciseResource.java [hunk 9]
```diff
@@ -468,7 +470,7 @@ public ProgrammingExerciseResource(ProgrammingExerciseRepository programmingExer
      * @param id the id of the programmingExercise to retrieve
      * @return the ResponseEntity with status 200 (OK) and with body the programmingExercise, or with status 404 (Not Found)
      */
-    @GetMapping("/programming-exercises/{id}")
+    @GetMapping(Endpoints.PROGRAMMING_EXERCISE)
     @PreAuthorize("hasAnyRole('TA', 'INSTRUCTOR', 'ADMIN')")
     public ResponseEntity<ProgrammingExercise> getProgrammingExercise(@PathVariable Long id) {
```
Comment (krusche) [inline src/main/java/de/tum/in/www1/artemis/web/rest/ProgrammingExerciseResource.java]:
```suggestion
    public ResponseEntity<ProgrammingExercise> getProgrammingExercise(@PathVariable Long exerciseId) {
```

File: src/main/java/de/tum/in/www1/artemis/web/rest/ProgrammingExerciseResource.java [hunk 10]
```diff
@@ -416,7 +418,7 @@ public ProgrammingExerciseResource(ProgrammingExerciseRepository programmingExer
      * @param notificationText to notify the student group about the updated problemStatement on the programming exercise
      * @return the ResponseEntity with status 200 (OK) and with body the updated problemStatement, with status 404 if the programmingExercise could not be found, or with 403 if the user does not have permissions to access the programming exercise.
      */
-    @PatchMapping("/programming-exercises-problem")
+    @PatchMapping(Endpoints.PROBLEM)
```
Comment (krusche) [inline src/main/java/de/tum/in/www1/artemis/web/rest/ProgrammingExerciseResource.java]: Ideally we refactor this URL to something like 

/programming-exercises/{exerciseId}/problem-statement

Comment (ungaralex) [reply src/main/java/de/tum/in/www1/artemis/web/rest/ProgrammingExerciseResource.java]: I refactored the endpoint and removed the now obsolete DTO `ProblemStatementUpdate`. I also updated the client and the related server test

**Pull request merged**